lib/faraday/multipart/middleware.rb

```diff
@@ -28,15 +27,46 @@ def call(env)
         @app.call env
       end
 
+      private
+
+      # @param env [Faraday::Env]
+      # @yield [request_body] Body of the request
+      def match_content_type(env)
+        return unless process_request?(env)
+
+        env.request_headers[CONTENT_TYPE] ||= mime_type
+        return if env.body.respond_to?(:to_str) || env.body.respond_to?(:read)
+
+        yield(env.body)
+      end
+
+      # @param env [Faraday::Env]
+      #
+      # @return [Boolean] True if the request has a body and its Content-Type is
+      #                   urlencoded.
+      def process_request?(env)
+        type = request_type(env)
+        env.body && (type.empty? || (type == mime_type))
+      end
+
       # @param env [Faraday::Env]
       def process_request?(env)
         type = request_type(env)
         env.body.respond_to?(:each_key) && !env.body.empty? && (
           (type.empty? && has_multipart?(env.body)) ||
-            (type == self.class.mime_type)
+            (type == mime_type)
         )
       end
 
+      # @param env [Faraday::Env]
+      #
+      # @return [String]
+      def request_type(env)
+        type = env.request_headers[CONTENT_TYPE].to_s
+        type = type.split(';', 2).first if type.index(';')
```

Comment:
Unrelated to this PR: I'll try fixing the "using index to check if the string contains a character" minor issue. There are other methods, like [`String#include?`](https://rubyapi.org/3.3/o/string#method-i-include-3F). Let's keep this code "as copied over".